clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java

```diff
@@ -1237,6 +1238,34 @@ public void testInitTransactionWhileThrottled() {
         }
     }
 
+    @Test
+    public void testClusterAuthorizationFailure() throws Exception {
+        int maxBlockMs = 500;
+
+        Map<String, Object> configs = new HashMap<>();
+        configs.put(ProducerConfig.MAX_BLOCK_MS_CONFIG, maxBlockMs);
+        configs.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9000");
+        configs.put(ProducerConfig.ENABLE_IDEMPOTENCE_CONFIG, true);
+        configs.put(ProducerConfig.TRANSACTIONAL_ID_CONFIG, "some-txn");
+
+        Time time = new MockTime(1);
+        MetadataResponse initialUpdateResponse = RequestTestUtils.metadataUpdateWith(1, singletonMap("topic", 1));
+        ProducerMetadata metadata = newMetadata(500, Long.MAX_VALUE);
+
+        MockClient client = new MockClient(time, metadata);
+        client.updateMetadata(initialUpdateResponse);
+
+        client.prepareResponse(FindCoordinatorResponse.prepareResponse(Errors.NONE, "some-txn", NODE));
+        client.prepareResponse(initProducerIdResponse(1L, (short) 5, Errors.CLUSTER_AUTHORIZATION_FAILED));
+        Producer<String, String> producer = kafkaProducer(configs, new StringSerializer(),
+                new StringSerializer(), metadata, client, null, time);
+        assertThrows(ClusterAuthorizationException.class, producer::initTransactions);
+
+        // retry initTransactions after the ClusterAuthorizationException not being thrown
+        client.prepareResponse(initProducerIdResponse(1L, (short) 5, Errors.NONE));
+        TestUtils.retryOnExceptionWithTimeout(1000, 100, producer::initTransactions);
```

Comment:
To be clear -- since I couldn't tell fully from the description -- before this change, trying to call initTransactions here would fail since the producer hits a fatal state. But with this change -- we don't retry initProducerId but instead keep the producer alive and basically re-initialize it so subsequent initTransactions calls can be made (and succeed).

Reply:
yes!

Reply:
Are we fully validating the request is successful? Is that what the TestUtils method does here?

Reply:
right, it retries initTransaction until successful or timeout.

Reply:
Should we close the producer here? Just looking through the failed tests and wanted to close any gaps we may have.

Reply:
Probably - it's just done very consistently in the tests 😅